cpp/dolfinx/mesh/utils.cpp

```diff
@@ -725,137 +726,190 @@ Eigen::Array<double, Eigen::Dynamic, 3, Eigen::RowMajor> mesh::midpoints(
-    // only for those
-    const std::vector<bool> on_boundary0 = mesh.topology().on_boundary(0);
-    for (std::size_t i = 0; i < on_boundary0.size(); ++i)
+    // Iterate over entity vertices
+    bool all_vertices_marked = true;
+    auto vertices = e_to_v->links(e);
+    for (int i = 0; i < vertices.rows(); ++i)
     {
-      if (on_boundary0[i])
-        active_vertex[i] = count++;
+      if (!marked[vertices[i]])
+      {
+        all_vertices_marked = false;
+        break;
+      }
     }
+
+    if (all_vertices_marked)
+      entities.push_back(e);
   }
-  else
+
+  return Eigen::Map<Eigen::Array<std::int32_t, Eigen::Dynamic, 1>>(
+      entities.data(), entities.size());
+}
+//-----------------------------------------------------------------------------
+Eigen::Array<std::int32_t, Eigen::Dynamic, 1> mesh::locate_entities_boundary(
+    const mesh::Mesh& mesh, const int dim,
+    const std::function<Eigen::Array<bool, Eigen::Dynamic, 1>(
+        const Eigen::Ref<const Eigen::Array<double, 3, Eigen::Dynamic,
+                                            Eigen::RowMajor>>&)>& marker)
+{
+  const mesh::Topology& topology = mesh.topology();
+  const int tdim = topology.dim();
+  if (dim == tdim)
   {
-    // Otherwise all vertices are active (will be checked with marking
-    // function)
-    std::iota(active_vertex.begin(), active_vertex.end(), 0);
-    count = num_vertices;
+    throw std::runtime_error(
+        "Cannot use mesh::locate_entities_boundary (boundary) for cells.");
   }
 
-  // FIXME: Does this make sense for non-affine elements?
-  // Get all nodes
-  const graph::AdjacencyList<std::int32_t>& x_dofmap = mesh.geometry().dofmap();
-  const Eigen::Array<double, Eigen::Dynamic, 3, Eigen::RowMajor>& x_all
-      = mesh.geometry().x();
-  auto v_to_c = mesh.topology().connectivity(0, tdim);
-  assert(v_to_c);
-  auto c_to_v = mesh.topology().connectivity(tdim, 0);
-  assert(c_to_v);
+  // Compute marker for boundary facets
+  mesh.topology_mutable().create_entities(tdim - 1);
+  mesh.topology_mutable().create_connectivity(tdim - 1, tdim);
+  const std::vector<bool> boundary_facet
+      = mesh::compute_boundary_facets(topology);
 
-  // Pack coordinates of all active vertices
-  Eigen::Array<double, 3, Eigen::Dynamic, Eigen::RowMajor> x_active(3, count);
+  // Create entities and connectivities
+  mesh.topology_mutable().create_entities(dim);
+  mesh.topology_mutable().create_connectivity(tdim - 1, dim);
+  mesh.topology_mutable().create_connectivity(tdim - 1, 0);
+  mesh.topology_mutable().create_connectivity(0, tdim);
+  mesh.topology_mutable().create_connectivity(tdim, 0);
+  if (dim < tdim)
+    mesh.topology_mutable().create_connectivity(dim, 0);
 
-  for (std::int32_t i = 0; i < num_vertices; ++i)
+  // Build set of vertices on boundary and set of boundary entities
+  auto f_to_v = topology.connectivity(tdim - 1, 0);
+  assert(f_to_v);
+  auto f_to_e = topology.connectivity(tdim - 1, dim);
```

Comment:
I guess in it should be possible to avoid some duplication here if dim == 0, but I am not sure if it is worth it or not.

Reply:
It could be optimised for `dim == 0` and `dim == tdim - 1`, but I've made it generic for now. `dim == tdim - 1` is perhaps the most common so could be looked at later.